manager_utils/upsert.py

```diff
@@ -0,0 +1,224 @@
+        for update_field in update_fields
+    ]
+
+    unique_field_names_sql = ', '.join([
+        _quote(field.column) for field in unique_fields
+    ])
+    update_fields_sql = ', '.join([
+        '{0} = EXCLUDED.{0}'.format(_quote(field.column))
+        for field in update_fields
+    ])
+
+    row_values = []
+    sql_args = []
+
+    for model_obj in model_objs:
+        placeholders = []
+        for field in all_fields:
+            # Convert field value to db value
+            # Use attname here to support fields with custom db_column names
+            sql_args.append(field.get_db_prep_save(getattr(model_obj, field.attname),
+                                                   connection))
+            placeholders.append('%s')
+        row_values.append('({0})'.format(', '.join(placeholders)))
+    row_values_sql = ', '.join(row_values)
+
+    return_sql = ''
+    if returning:
+        action_sql = ', (xmax = 0) AS inserted_'
+        if returning is True:
+            return_sql = 'RETURNING * {action_sql}'.format(action_sql=action_sql)
+        else:
+            return_fields_sql = ', '.join(_quote(field) for field in returning)
+            return_sql = 'RETURNING {return_fields_sql} {action_sql}'.format(return_fields_sql=return_fields_sql,
+                                                                             action_sql=action_sql)
+
+    if update_fields:
+        sql = (
+            'INSERT INTO {0} ({1}) VALUES {2} ON CONFLICT ({3}) DO UPDATE SET {4} {5}'
+        ).format(
+            model._meta.db_table,
+            all_field_names_sql,
+            row_values_sql,
+            unique_field_names_sql,
+            update_fields_sql,
+            return_sql
+        )
+    else:
+        sql = (
+            'INSERT INTO {0} ({1}) VALUES {2} ON CONFLICT ({3}) {4} {5}'
+        ).format(
+            model._meta.db_table,
+            all_field_names_sql,
+            row_values_sql,
+            unique_field_names_sql,
+            'DO UPDATE SET {0}=EXCLUDED.{0}'.format(unique_fields[0].column),
+            return_sql
+        )
+
+    return sql, sql_args
+
+
+def _fetch(queryset, model_objs, unique_fields, update_fields, returning, sync):
+    """Perfom the upsert and do an optional sync operation"""
+    model = queryset.model
+    upserted = []
+    deleted = []
+    if model_objs:
+        sql, sql_args = _get_upsert_sql(queryset, model_objs, unique_fields, update_fields, returning)
+
+        with connection.cursor() as cursor:
+            cursor.execute(sql, sql_args)
+            if cursor.description:
+                nt_result = namedtuple('Result', [col[0] for col in cursor.description])
+                upserted = [nt_result(*row) for row in cursor.fetchall()]
+
+    pk_field = model._meta.pk.name
+    if sync:
+        orig_ids = frozenset(queryset.values_list(pk_field, flat=True))
+        deleted = list(orig_ids - frozenset([getattr(r, pk_field) for r in upserted]))
```

Comment:
any reason to do a frozenset of a list rather than a set comprehension? something like ```{getattr(r, pk_field) for r in upserted}```

also, would it be better to leave the diff as a set rather than converting it to a list?

Reply:
@wesokes nope, looks like that was just some old code that was copied and pasted from a long time ago. I updated it to remove the unnecessary list at the end and did a set comprehension